### PR TITLE
deprecate surface::VTKUtils

### PR DESCRIPTION
**Pull request by greenbrettmichael**
Deprecate duplicate vtk conversion code used by vtk_smoothing in pcl surface in favor of conversion code in io.
This resolves an issue with the surface conversion not transferring some scalar properties of the data.
This add io as a dependency to surface if VTK_FOUND is true
closes https://github.com/PointCloudLibrary/pcl/issues/2646

### CODE REVIEW

**Comment by SergioRAgostinho** [comment]
Till I lay my eyes on this PR there's something you can have a look in the meantime. The new deprecations are producing  warnings during compilation.
```
2018-11-28T18:23:23.1183747Z /__w/1/s/surface/src/vtk_smoothing/vtk_utils.cpp: In static member function 'static int pcl::VTKUtils::convertToVTK(const pcl::PolygonMesh&, vtkSmartPointer<vtkPolyData>&)':
2018-11-28T18:23:23.1184639Z /__w/1/s/surface/src/vtk_smoothing/vtk_utils.cpp:56:3: warning: 'static int pcl::VTKUtils::mesh2vtk(const pcl::PolygonMesh&, vtkSmartPointer<vtkPolyData>&)' is deprecated: Use pcl::io::mesh2vtk instead [-Wdeprecated-declarations]
2018-11-28T18:23:23.1184727Z    mesh2vtk (triangles, vtk_polygons);
2018-11-28T18:23:23.1184816Z    ^
2018-11-28T18:23:23.1185423Z In file included from /__w/1/s/surface/src/vtk_smoothing/vtk_utils.cpp:40:0:
2018-11-28T18:23:23.1185482Z /__w/1/s/surface/include/pcl/surface/vtk_smoothing/vtk_utils.h:83:7: note: declared here
2018-11-28T18:23:23.1185556Z        mesh2vtk (const pcl::PolygonMesh& mesh, vtkSmartPointer<vtkPolyData> &poly_data);
2018-11-28T18:23:23.1185599Z        ^
2018-11-28T18:23:23.1189018Z /__w/1/s/surface/src/vtk_smoothing/vtk_utils.cpp:56:36: warning: 'static int pcl::VTKUtils::mesh2vtk(const pcl::PolygonMesh&, vtkSmartPointer<vtkPolyData>&)' is deprecated: Use pcl::io::mesh2vtk instead [-Wdeprecated-declarations]
2018-11-28T18:23:23.1189323Z    mesh2vtk (triangles, vtk_polygons);
2018-11-28T18:23:23.1189472Z                                     ^
2018-11-28T18:23:23.1190627Z In file included from /__w/1/s/surface/src/vtk_smoothing/vtk_utils.cpp:40:0:
2018-11-28T18:23:23.1190804Z /__w/1/s/surface/include/pcl/surface/vtk_smoothing/vtk_utils.h:83:7: note: declared here
2018-11-28T18:23:23.1191176Z        mesh2vtk (const pcl::PolygonMesh& mesh, vtkSmartPointer<vtkPolyData> &poly_data);
2018-11-28T18:23:23.1191356Z        ^
2018-11-28T18:23:23.1201464Z /__w/1/s/surface/src/vtk_smoothing/vtk_utils.cpp: In static member function 'static void pcl::VTKUtils::convertToPCL(vtkSmartPointer<vtkPolyData>&, pcl::PolygonMesh&)':
2018-11-28T18:23:23.1202254Z /__w/1/s/surface/src/vtk_smoothing/vtk_utils.cpp:75:3: warning: 'static int pcl::VTKUtils::vtk2mesh(const vtkSmartPointer<vtkPolyData>&, pcl::PolygonMesh&)' is deprecated: Use pcl::io::vtk2mesh instead [-Wdeprecated-declarations]
2018-11-28T18:23:23.1202395Z    vtk2mesh (vtk_polygons, triangles);
2018-11-28T18:23:23.1202443Z    ^
2018-11-28T18:23:23.1205249Z In file included from /__w/1/s/surface/src/vtk_smoothing/vtk_utils.cpp:40:0:
2018-11-28T18:23:23.1206110Z /__w/1/s/surface/include/pcl/surface/vtk_smoothing/vtk_utils.h:74:7: note: declared here
2018-11-28T18:23:23.1206252Z        vtk2mesh (const vtkSmartPointer<vtkPolyData>& poly_data, pcl::PolygonMesh& mesh);
2018-11-28T18:23:23.1206336Z        ^
2018-11-28T18:23:23.1210119Z /__w/1/s/surface/src/vtk_smoothing/vtk_utils.cpp:75:36: warning: 'static int pcl::VTKUtils::vtk2mesh(const vtkSmartPointer<vtkPolyData>&, pcl::PolygonMesh&)' is deprecated: Use pcl::io::vtk2mesh instead [-Wdeprecated-declarations]
2018-11-28T18:23:23.1210279Z    vtk2mesh (vtk_polygons, triangles);
2018-11-28T18:23:23.1210333Z                                     ^
2018-11-28T18:23:23.1213235Z In file included from /__w/1/s/surface/src/vtk_smoothing/vtk_utils.cpp:40:0:
2018-11-28T18:23:23.1213811Z /__w/1/s/surface/include/pcl/surface/vtk_smoothing/vtk_utils.h:74:7: note: declared here
2018-11-28T18:23:23.1214063Z        vtk2mesh (const vtkSmartPointer<vtkPolyData>& poly_data, pcl::PolygonMesh& mesh);
2018-11-28T18:23:23.1214134Z        ^
```
With this PR I expect us to stop using any of the functions which were deprecated.

File: surface/CMakeLists.txt [hunk 1]
```diff
@@ -45,9 +45,9 @@ if(build)
             src/vtk_smoothing/vtk_mesh_smoothing_windowed_sinc.cpp)
         
         if("${VTK_MAJOR_VERSION}.${VTK_MINOR_VERSION}" VERSION_LESS "6.0")
-            set(VTK_SMOOTHING_TARGET_LINK_LIBRARIES vtkCommon vtkWidgets vtkGraphics)
+            set(VTK_SMOOTHING_TARGET_LINK_LIBRARIES pcl_io vtkCommon vtkWidgets vtkGraphics)
```
Comment (SergioRAgostinho) [inline surface/CMakeLists.txt]: Before adding this here, we need to report to `pcl_io` also as an optional dependency. The check now needs to be "bumped"  from simply finding VTK, to now ensuring `pcl_io` is built and also checking VTK is present.

Comment (SergioRAgostinho) [reply surface/CMakeLists.txt]:
```diff
diff --git a/surface/CMakeLists.txt b/surface/CMakeLists.txt
index 719e88d41..1264fb631 100644
--- a/surface/CMakeLists.txt
+++ b/surface/CMakeLists.txt
@@ -1,10 +1,11 @@
 set(SUBSYS_NAME surface)
 set(SUBSYS_DESC "Point cloud surface library")
 set(SUBSYS_DEPS common search kdtree octree)
+set(SUBSYS_OPT_DEPS io qhull)
 
 set(build TRUE)
 PCL_SUBSYS_OPTION(build "${SUBSYS_NAME}" "${SUBSYS_DESC}" ON)
-PCL_SUBSYS_DEPEND(build "${SUBSYS_NAME}" DEPS ${SUBSYS_DEPS} OPT_DEPS qhull)
+PCL_SUBSYS_DEPEND(build "${SUBSYS_NAME}" DEPS ${SUBSYS_DEPS} OPT_DEPS ${SUBSYS_OPT_DEPS})
 
 PCL_ADD_DOC("${SUBSYS_NAME}")
 
@@ -24,8 +25,9 @@ if(build)
             src/convex_hull.cpp
             )
     endif(QHULL_FOUND)
-    
-    if (VTK_FOUND AND NOT ANDROID)
+
+    PCL_GET_SUBSYS_STATUS(IO_STATUS io)
+    if (IO_STATUS AND VTK_FOUND AND NOT ANDROID)
         set(VTK_USE_FILE "${VTK_USE_FILE}" CACHE INTERNAL "VTK_USE_FILE")
         include("${VTK_USE_FILE}")
         
   
```

**Comment by greenbrettmichael** [comment]
@SergioRAgostinho the feedback has been addressed

**Comment by taketwo** [comment]
Guys, I'd like to also have a deeper look at this PR before merging. This might take a few days before I get to this, sorry for holding back.

File: surface/CMakeLists.txt [hunk 2]
```diff
@@ -45,11 +47,11 @@ if(build)
...
         else("${VTK_MAJOR_VERSION}.${VTK_MINOR_VERSION}" VERSION_LESS "6.0")
-            set(VTK_SMOOTHING_TARGET_LINK_LIBRARIES vtkCommonCore vtkCommonDataModel vtkCommonExecutionModel vtkFiltersModeling)
+            set(VTK_SMOOTHING_TARGET_LINK_LIBRARIES pcl_io vtkCommonCore vtkCommonDataModel vtkCommonExecutionModel vtkFiltersModeling)
         endif("${VTK_MAJOR_VERSION}.${VTK_MINOR_VERSION}" VERSION_LESS "6.0")
-    endif()
+    endif(IO_STATUS AND VTK_FOUND AND NOT ANDROID)
```
Comment (SergioRAgostinho) [inline surface/CMakeLists.txt]: One final detail from my side. We're no longer applying this pattern

```cmake
if(CONDITION)
else(CONDITION)
endif(CONDITION)
```
This was required in older cmake versions this practice was dropped.
```suggestion
    endif()
```

File: surface/CMakeLists.txt [hunk 3]
```diff
@@ -175,9 +177,9 @@ if(build)
       PCL_ADD_INCLUDES("${SUBSYS_NAME}" "${SUBSYS_NAME}/on_nurbs" ${ON_NURBS_INCLUDES})
     endif(BUILD_surface_on_nurbs)
 
-    if (VTK_FOUND AND NOT ANDROID)
+    if (IO_STATUS AND VTK_FOUND AND NOT ANDROID)
       PCL_ADD_INCLUDES("${SUBSYS_NAME}" "${SUBSYS_NAME}/vtk_smoothing" ${VTK_SMOOTHING_INCLUDES})
-    endif (VTK_FOUND AND NOT ANDROID)
+    endif (IO_STATUS AND VTK_FOUND AND NOT ANDROID)
```
Comment (SergioRAgostinho) [inline surface/CMakeLists.txt]: Same here
```suggestion
    endif ()
```

**Review by SergioRAgostinho — DISMISSED** [review]
Thanks 👍

**Comment by taketwo** [comment]
:+1: for what you guys decided to do, makes sense to me. While we are at this, what about merging `vkt_lib_io.h` into `vtk_io.h`? I don't see logic behind separating these.

**Comment by SergioRAgostinho** [comment]
Instead of merging the latest changes from master, you need to rebase your work on top of pcl master.

**Comment by taketwo** [comment]
Please note that In the meantime we started using `[[deprecated]]` attribute (#2699) and increased minimum required VTK version to 6.2 (#2685).

**Comment by greenbrettmichael** [comment]
@SergioRAgostinho applied the rebase instead, @taketwo merged io and lib_io , this could not be done with deprecation since both declarations were in pcl::io and would result in multiply defined (keeping impl) or unresolved symbols (removing impl). This means this PR is now a breaking change if you use vtk_lib_io.h

**Comment by taketwo** [comment]
Thanks. Two points from my side:

1. Please add short strings explaining which function should be used instead of deprecated, e.g.:

```cpp
[[deprecated("use mesh2vtk() instead]]
```

2. We shouldn't entirely remove `vtk_lib_io.h/hpp` files to not break user projects. We need to keep them around for some time so that people can make a transition. The files should be empty except to a pragma warning that they are deprecated and in include of the corresponding `vtk_io.h/hpp` file instead.

I'm not sure I understand your last message, but perhaps my second point is related.

**Comment by greenbrettmichael** [comment]
@taketwo I understand point 2 but do not see how it can be done without producing its own breaking change (compilation errors from multiple definition). I have pushed a commit with what I believe to be the intended result.

**Comment by SergioRAgostinho** [comment]
Similar to what was done [here](https://github.com/PointCloudLibrary/pcl/pull/164/files#diff-1d4dad08d20f16064c04ccecf40910fd). A number of conversion function were moved to `pcl/conversions.h`. `pcl/ros/conversions.h` throws the warning and then includes `pcl/conversions.h`, to ensure that functions if previously had, are still visible to the user.

File: io/include/pcl/io/vtk_lib_io.h [hunk 4]
```diff
@@ -1,271 +1,48 @@
...
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *
+ * $Id$
+ *
+ */
+
+#pragma once
+
+#ifdef __DEPRECATED
```
Comment (SergioRAgostinho) [inline io/include/pcl/io/vtk_lib_io.h]: @taketwo any idea if we actually need this check?

File: io/include/pcl/io/vtk_io.h [hunk 5]
```diff
@@ -68,5 +95,196 @@ namespace pcl
...
     PCL_EXPORTS int 
     saveVTKFile (const std::string &file_name, const pcl::PCLPointCloud2 &cloud, unsigned precision = 5);
+    /** \brief Convert vtkPolyData object to a PCL PolygonMesh
+      * \param[in] poly_data Pointer (vtkSmartPointer) to a vtkPolyData object
+      * \param[out] mesh PCL Polygon Mesh to fill
+      * \return Number of points in the point cloud of mesh.
+      */
+    PCL_EXPORTS int
+      vtk2mesh(const vtkSmartPointer<vtkPolyData>& poly_data,
+        pcl::PolygonMesh& mesh);
```
[the remark this answers is not in the document]
Comment (greenbrettmichael) [reply io/include/pcl/io/vtk_io.h]: It appears my IDE mangled these on paste, I have disabled most of the formatting options but having a clang format file would really help with this.

**Comment by kunaltyagi** [comment]
This PR is almost finished. If the OP or anyone else wishes to take over, that'd be nice.

**Comment by greenbrettmichael** [comment]
I am not going to be able to come back to this (I barely remember the changes). Closing the PR.

**Pull request closed without merging**